rfd/0109-opsgenie-plugin.md

```diff
@@ -0,0 +1,169 @@
+# RFD 109: Opsgenie Integration
+## Required approvers
+
+Engineering: @r0mant, @marcoandredinis, @hugoShaka
+Product: @klizhentas, @xinding33
+Security: @reedloden, @jentfoo
+
+## What
+
+This RFD proposes a plugin that allows Teleport to integrate with Opsgenie, allowing access requests to show up as alerts in Opsgenie. This plugin will differ from existing plugins by being part of the Teleport binary directly. 
+Doing so by creating a new Teleport service 'plugin_service' with the opsgenie plugin (and any future ones) being configurable resources. This will allow the plugins to be run either by themselves or alongside other Teleport services.
+
+## Scope
+
+The plugin will be able to run standalone or alongside other services as part of the new plugin_service.
+The plugin will support auto-approval similar to PagerDuty plugin.
+
+## Success criteria
+
+### Plugin service
+The creation of a Teleport service where plugins can be created as configurable resources similar to the existing 'app_service' and 'db_service'.
+
+Plugins managed through this service should have the ability to be enabled and disabled dynamically via 'tctl'.
+
+### Opsgenie plugin
+Users are able to configure teleport to automatically create alerts in Opsgenie from access requests.
+Alerts were chosen over incidents in Opsgenie as incidents are intended to be used for high priority alerts that indicate a service interruption.
+Users are also able to configure auto approval flows to be met under certain conditions. E.g when a requester is on-call.
+
+## Configuration UX
+
+The plugin service will be configured using the existing Teleport YAML file in a section called 'plugin_service'.
+
+```
+plugin_service:
+    enabled: true
+    plugins:
+    - "type": "opsgenie"
+    - "type": "pagerduty"
+    opsgenie:
+        api_key: "path/to/key.txt" # Path to a file containing the Opsgenie API key
+```
+This example would only match plugin resources with the the labels 'type:opsgenie', or 'type:slack'.
+
+The Opsgenie plugin (and any others created) can then be configured using resources. 
+Example plugin.yaml for Opsgenie that would match with this.
+```
+kind: plugin
+metadata:
+  name: opsgenie-plugin
+  labels:
+    type: opsgenie
+spec:
+  opsgenie:
```

Comment:
Can one run several OpsGenie plugins using different `api_key`-s? If so, it should be possible to specify the auth secret (API key) in the Plugin resource (Plugin.[credentials](https://github.com/gravitational/teleport/blob/1c3188a38a312d7f85153d172ed52a1dfb0ad996/api/proto/teleport/legacy/types/types.proto#L4921-L4923)) rather than globally (?) in `plugins_service` config.

In any case, this would be a prerequisite for OpsGenie plugin in Cloud, as the users would need to supply the API key in some way, and they can not edit the config in Cloud.

Reply:
At the moment was not considering multiple API keys. We wanted to avoid putting secrets in the resource.

Reply:
Yeah, we'd like to avoid secrets in the resource. This would likely means that we won't be able to support OpsGenie in the Cloud in the fully-hosted way as users will have to deploy an agent that has an API key but that's fine for now (it was discussed with the product team).